File: core/domain/src/commonTest/kotlin/com/syrmos/core/domain/journey/JourneyDetailTest.kt
package com.syrmos.core.domain.journey

import com.syrmos.core.model.journey.Feasibility
import com.syrmos.core.model.journey.FeasibilityStatus
import com.syrmos.core.model.journey.JourneyOption
import com.syrmos.core.model.journey.Leg
import com.syrmos.core.model.journey.LegKind
import com.syrmos.core.model.journey.TimingKind
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDate
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Kotlin peer of web `web-tests/journey-detail.test.js`, mirroring the golden
 * cases in fixtures/journeys/detail.json. Proves the S05 timeline transform emits
 * the same ordered rows on Kotlin as on web/iOS (node roles, intermediate-stop
 * counts, transfer-gap fallback, and null-clock honesty).
 */
class JourneyDetailTest {
    private val date = LocalDate.parse("2026-01-15")
    private fun inst(s: String) = Instant.parse(s)

    private fun ride(id: String, line: String, from: String, to: String, stops: List<String>,
                     dep: String?, arr: String?, timing: TimingKind = TimingKind.SCHEDULED) =
        Leg(id = id, kind = LegKind.RIDE, fromId = from, toId = to, lineId = line,
            orderedStopIds = stops, departureInstant = dep?.let(::inst), arrivalInstant = arr?.let(::inst),
            serviceDate = date, timingKind = timing)

    private fun transfer(id: String, from: String, to: String, minSeconds: Int? = null, walk: Boolean = false) =
        Leg(id = id, kind = if (walk) LegKind.WALK else LegKind.TRANSFER, fromId = from, toId = to,
            orderedStopIds = listOf(from, to), serviceDate = date, timingKind = TimingKind.ESTIMATED,
            transferMinimumSeconds = minSeconds)

    private fun option(vararg legs: Leg) = JourneyOption(
        id = "o", requestId = "r", legs = legs.toList(),
        transferCount = legs.count { it.kind == LegKind.RIDE } - 1,
        feasibility = Feasibility(FeasibilityStatus.UNKNOWN, explanationCode = "pending"),
    )

    @Test
    fun twoRidesOneTransfer() {
        val opt = option(
            ride("ride-0", "M1", "M1_PIR", "M1_MON", listOf("M1_PIR", "M1_FAL", "M1_TAV", "M1_MON"),
                "2026-01-15T08:00:00+02:00", "2026-01-15T08:15:00+02:00"),
            transfer("transfer-1", "M1_MON", "M3_MON", minSeconds = 180),
            ride("ride-2", "M3", "M3_MON", "M3_SYN", listOf("M3_MON", "M3_SYN"),
                "2026-01-15T08:20:00+02:00", "2026-01-15T08:25:00+02:00"),
        )
        val rows = JourneyDetail.timeline(opt)
        assertEquals(
            listOf("board", "stops", "alight", "transfer", "board", "alight"),
            rows.map { it.kind },
        )
        assertEquals(
            listOf(JourneyDetail.Node.ORIGIN, JourneyDetail.Node.INTERCHANGE, JourneyDetail.Node.INTERCHANGE, JourneyDetail.Node.INTERCHANGE, JourneyDetail.Node.DESTINATION),
            rows.mapNotNull { it.node },
        )
        val board0 = rows[0]
        assertEquals("M1", board0.lineId); assertEquals("M1_MON", board0.towardsId)
        assertEquals(inst("2026-01-15T08:00:00+02:00"), board0.clock)
        assertEquals("scheduled", board0.timingKind)
        assertEquals(2, rows[1].count)
        assertEquals(180, rows[3].seconds)
    }

    @Test
    fun directRideKeepsNullClocks() {
        val opt = option(
            ride("ride-0", "M2", "M2_SYN", "M2_ELL", listOf("M2_SYN", "M2_SYG", "M2_ELL"),
                null, null, timing = TimingKind.ESTIMATED),
        )
        val rows = JourneyDetail.timeline(opt)
        assertEquals(listOf("board", "stops", "alight"), rows.map { it.kind })
        assertNull(rows.first().clock)
        assertNull(rows.last().clock)
        assertEquals(JourneyDetail.Node.ORIGIN, rows.first().node)
        assertEquals(JourneyDetail.Node.DESTINATION, rows.last().node)
        assertEquals(1, rows[1].count)
    }

    @Test
    fun walkTransferFallsBackToGap() {
        val opt = option(
            ride("ride-0", "M1", "M1_PIR", "M1_MON", listOf("M1_PIR", "M1_MON"),
                "2026-01-15T08:05:00+02:00", "2026-01-15T08:15:00+02:00"),
            transfer("walk-1", "M1_MON", "M3_MON", walk = true),
            ride("ride-2", "M3", "M3_MON", "M3_SYN", listOf("M3_MON", "M3_SYN"),
                "2026-01-15T08:20:00+02:00", "2026-01-15T08:24:00+02:00"),
        )
        val rows = JourneyDetail.timeline(opt)
        assertTrue(rows.none { it.kind == "stops" }, "no stops disclosure for 2-stop legs")
        val walk = rows.first { it.kind == "walk" }
        assertEquals(300, walk.seconds)
    }
}

File: web-tests/journey-detail.test.js
'use strict';
// 3.0 S05: the web selected-journey detail timeline transform must match the
// shared golden fixture fixtures/journeys/detail.json exactly — the same ordered
// rows Kotlin JourneyDetail.timeline and iOS JourneyDetail.timeline produce. This
// is the cross-client parity guard for the S05 timeline.
const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');

const RES = path.join(__dirname, '..', 'composeApp', 'src', 'wasmJsMain', 'resources');
const FIX = path.join(__dirname, '..', 'fixtures', 'journeys', 'detail.json');
const Detail = require(path.join(RES, 'web-journey-detail.js'));
const fixture = JSON.parse(fs.readFileSync(FIX, 'utf8'));

test('every fixture case produces the exact expected timeline rows', () => {
  for (const c of fixture.cases) {
    const rows = Detail.timeline(c.option);
    assert.deepEqual(rows, c.expectedRows, c.id);
  }
});

test('a null clock stays null (unknown), never a fabricated value', () => {
  const rows = Detail.timeline({ legs: [
    { id: 'r0', kind: 'ride', lineId: 'M2', fromId: 'A', toId: 'B', orderedStopIds: ['A', 'B'], departureInstant: null, arrivalInstant: null, timingKind: 'estimated' },
  ] });
  assert.equal(rows[0].kind, 'board');
  assert.equal(rows[0].clock, null);
  assert.equal(rows[1].kind, 'alight');
  assert.equal(rows[1].clock, null);
});

test('origin/destination/interchange node roles are assigned by position', () => {
  const rows = fixture.cases.find((c) => c.id === 'two_rides_one_transfer').expectedRows;
  assert.equal(rows.find((r) => r.kind === 'board' && r.legId === 'ride-0').node, 'origin');
  assert.equal(rows.find((r) => r.kind === 'alight' && r.legId === 'ride-0').node, 'interchange');
  assert.equal(rows.find((r) => r.kind === 'alight' && r.legId === 'ride-2').node, 'destination');
  // recompute from the transform, not just the fixture literal
  const live = Detail.timeline(fixture.cases.find((c) => c.id === 'two_rides_one_transfer').option);
  assert.deepEqual(live.map((r) => r.node).filter(Boolean), ['origin', 'interchange', 'interchange', 'interchange', 'destination']);
});

test('intermediate stops are counted, and no stops row when <= 2 ordered stops', () => {
  const rows = Detail.timeline({ legs: [
    { id: 'r0', kind: 'ride', lineId: 'M1', fromId: 'A', toId: 'D', orderedStopIds: ['A', 'B', 'C', 'D'], timingKind: 'scheduled' },
  ] });
  const stops = rows.find((r) => r.kind === 'stops');
  assert.ok(stops && stops.count === 2, 'two intermediate stops');
  const rows2 = Detail.timeline({ legs: [
    { id: 'r0', kind: 'ride', lineId: 'M1', fromId: 'A', toId: 'B', orderedStopIds: ['A', 'B'], timingKind: 'scheduled' },
  ] });
  assert.ok(!rows2.some((r) => r.kind === 'stops'), 'no stops disclosure for a 2-stop leg');
});

test('transfer seconds fall back to the arrival->departure gap when no minimum is set', () => {
  const c = fixture.cases.find((x) => x.id === 'walk_transfer_gap_seconds');
  const rows = Detail.timeline(c.option);
  const walk = rows.find((r) => r.kind === 'walk');
  assert.equal(walk.seconds, 300);
});
